backend/student_on_tour/views.py

```diff
@@ -31,6 +30,151 @@ def post(self, request):
         return post_success(StudOnTourSerializer(student_on_tour_instance))
 
 
+class StudentOnTourBulk(APIView):
+    permission_classes = [IsAuthenticated, IsAdmin | IsSuperStudent]
+    serializer_class = StudOnTourSerializer
+
+    @extend_schema(
+        description="POST body consists of a data component that is a list of Student-Tour instances. "
+        "This enables the frontend to save a schedule in 1 request instead of multiple. "
+        "If a save fails, all the previous saves will be undone as well.",
+        request=StudOnTourSerializer,
+        responses={200: SuccessSerializer, 400: None},
+        examples=[
+            OpenApiExample(
+                "Request body for bulk add",
+                value={
+                    "data": [
+                        {"tour": 0, "student": 3, "date": "2023-04-28"},
+                        {"tour": 1, "student": 2, "date": "2023-04-28"},
+                    ]
+                },
+                description="",
+                request_only=True,
+            )
+        ],
+    )
+    def post(self, request):
+        data = request_to_dict(request.data)
+        """
+        request body should look like this:
+        {
+            data:
+            [
+                {Tour:x, student:x, date: x},
+                {Tour:x2, student:x, date: x2},
+                // more of this
+            ]
+        }
+        """
+        list_done = []
+        for d in data["data"]:
+            print(d)
+            student_on_tour_instance = StudentOnTour()
+
+            set_keys_of_instance(student_on_tour_instance, d, TRANSLATE)
+
+            if r := try_full_clean_and_save(student_on_tour_instance):
+                for elem in list_done:
+                    elem.delete()
+                return r
+            list_done.append(student_on_tour_instance)
```

Comment:
Why are you saving the saving the student_on_tour_instances in this list_done?

Reply:
It's a history of the previous saves, so if 1 fails along the way, I can remove all the previous ones as well.